### PR TITLE
fix(predictor): add balls-in-over slider for per-ball granularity

### LINKED ISSUE #155: [BUG] balls_left computed from whole overs only — model trained with ball-level precision but inference always assumes ball=0 of each over, producing wrong RRR for every mid-over prediction
## Describe the Bug

The win probability model is trained with per-ball granularity but the prediction UI only accepts a whole-over input. This creates a persistent feature mismatch that produces systematically incorrect `balls_left` and `rrr` values for every prediction made mid-over.

### Training code (ball-level precision)

**File:** `application.py`

```python
# Training — uses both over AND ball within that over
df['balls_left'] = 120 - (df['over'] * 6 + df['ball'])
df['crr']        = df['current_score'] / (df['over'] + df['ball'] / 6)
df['rrr']        = (df['runs_left'] * 6) / df['balls_left']
```

### Prediction code (coarse over-level only)

**File:** `application.py`

```python
# Prediction — ball column is completely absent
balls_left = 120 - (overs * 6)       # always assumes ball = 0 of the over
crr        = score / overs            # always divides by whole overs
rrr        = (runs_left * 6) / balls_left if balls_left > 0 else 0
```

### Concrete example

A match is at **over 15, ball 3** (i.e., 15.3 overs). The true state:

| Feature | Training value | Prediction value | Error |
|---|---|---|---|
| `balls_left` | 120 - (15*6 + 3) = **57** | 120 - (15*6) = **60** | **+3 balls** |
| `crr` (score=120) | 120 / (15 + 3/6) = **7.74** | 120 / 15 = **8.00** | **+0.26** |
| `rrr` (runs_left=60) | (60*6) / 57 = **6.32** | (60*6) / 60 = **6.00** | **-0.32** |

The model is told there are 3 more balls remaining than there actually are. The effective run rates are wrong in opposite directions, compounding the prediction error. This happens on every single prediction unless the user happens to be exactly at the start of an over.

## Steps to Reproduce

1. Open the Analysis page
2. Set Overs to 10 (slider selects whole overs only — no sub-over field)
3. Run a prediction at e.g. 10 overs, 80 runs, target 160
4. The model receives `balls_left=60` but during training, ball 10.3 would give `balls_left=57`

## Expected Behavior

Predictions should accept a "balls bowled in current over" input (0-5) and derive:

```python
balls_left = 120 - (overs * 6 + balls_in_over)
crr        = score / (overs + balls_in_over / 6) if (overs + balls_in_over / 6) > 0 else 0
rrr        = (runs_left * 6) / balls_left if balls_left > 0 else 0
```

## Suggested Fix

Add a `st.slider("Balls in Current Over", min_value=0, max_value=5, value=0, step=1, key="balls_in_over")` widget alongside the overs input. Wire it into the `balls_left` and `crr` derivations to match training-time feature engineering exactly.

## Impact

Every win probability prediction made at a non-zero ball within an over is computed from incorrect feature values. Because the over slider forces integer overs, **all real-world mid-over queries are affected.** The error grows larger in the final overs where individual balls matter most for run rate calculations.

## Environment

- File: `application.py`
- Lines: training ~655-664, prediction ~997-1002
- Model: XGBoost via sklearn Pipeline

### LINKED ISSUE #156: [BUG] Overs slider hard-capped at max_value=19 — 20th-over (final over) win probability predictions are completely unreachable from the UI
## Describe the Bug

The overs slider on the Analysis page has `max_value=19`, which makes it impossible to generate a win probability prediction for the last over of a T20 match. T20 cricket has 20 overs per innings, and the final over is frequently the most critical moment for prediction. The model was trained on last-over data (over 20), but the UI blocks access to it.

**File:** `application.py`

```python
overs = st.slider("Overs Completed", min_value=1, max_value=19, value=10)
```

## Steps to Reproduce

1. Open the Analysis page
2. Drag the "Overs Completed" slider to its maximum position
3. Maximum is 19, not 20
4. No way to request a prediction for ball 1-6 of the final over

## Expected Behavior

The slider maximum should be `20` to allow prediction across the full innings, including the final over. Alternatively, if the intent is "overs completed" (i.e., full overs bowled), the max should be 19 with clear labeling, but a separate "balls in current over" input (0-5) should allow representing the last over's progression — which ties directly into the `balls_left` feature.

## Fix

```python
overs = st.slider("Overs Completed", min_value=1, max_value=20, value=10)
```

Additionally, pair this with a bounds check so `balls_left` does not go negative when `overs=20`:

```python
balls_left = max(0, 120 - (overs * 6))
```

## Impact

Any tense chase scenario in the 20th over — which accounts for a significant share of IPL finishes — cannot be analyzed through the UI at all. Users attempting to model last-over scenarios get no result and no error message, making the tool incomplete for its primary use case.

## Environment

- File: `application.py`
- Line: ~911
- Affects: T20 win probability prediction page

### PR DESCRIPTION
**Pull request by anshul23102**
Fixes #155
Fixes #156

## Problem

The win-probability predictor tracked match progress only in whole overs, discarding up to 5 balls of precision for any delivery bowled mid-over. As a result:

- `balls_left` was always a multiple of 6, never reflecting the true number of balls remaining
- `crr` divided by integer overs, giving the wrong rate during an ongoing over
- `rrr` inherited the same inaccuracy

Additionally (issue #156), the overs slider had `max_value=19` in an earlier version; the calculation now uses `max(120 - total_balls_bowled, 0)` which correctly clamps the result even if the full 20 overs are entered.

## Fix

Added a **"Balls in Current Over"** slider (0-5) in the Match State panel and updated all three derived metrics:

```python
total_balls_bowled = overs * 6 + balls_in_over
balls_left = max(120 - total_balls_bowled, 0)
crr = score / (total_balls_bowled / 6) if total_balls_bowled > 0 else 0.0
rrr = (runs_left * 6) / balls_left if balls_left > 0 else 0.0
```

The slider sits between "Overs Completed" and "Wickets Fallen" in a three-column layout, keeping the input area compact.

## Testing

- Over 10, ball 3: `balls_left = 120 - 63 = 57` (was 60 before)
- Over 0, ball 0: `crr = 0.0` (zero-division guard still holds)
- Over 20, ball 0: `balls_left = 0`, model falls into the `balls_left <= 0` edge-case path

Could you please add the appropriate labels (`level:*`, `type:bug`, `gssoc'26`) when you get a chance? Thank you!

### CODE REVIEW

**Comment by anshul23102** [comment]
Hi @Arnav-Singh-5080, this PR is ready for review. No merge conflicts with main and the branch is rebased on the current main tip. Could you please add the relevant labels (bug, GSSoC '26) when reviewing? Thank you!

**Comment by anshul23102** [comment]
Hi @Arnav-Singh-5080, this PR had a merge conflict which I have now resolved via a rebase. The branch is clean and mergeable. Please let me know if you need anything further. Thank you!

**Comment by anshul23102** [comment]
Gentle ping -- this PR has been open for 3 days with no activity. Could you please take a look when you get a chance? Happy to address any feedback.

**Comment by anshul23102** [comment]
Hi @Arnav-Singh-5080, a polite follow-up on this PR. It has been a few days since any activity. All checks are passing and there are no merge conflicts. Please review when you have a chance. Happy to make any adjustments you need.

**Comment by anshul23102** [comment]
Hi @Arnav-Singh-5080, following up on this PR - it addresses both #155 (ball-level RRR precision) and #156 (overs slider cap) and has been open for a while awaiting review. Could you please take a look when you get a chance, and apply the appropriate GSSoC'26 labels if it looks good? Thanks for your time.
